### PR TITLE
Add Tents

### PR DESCRIPTION
**Pull request by x-sheep**
To do:
* ~Replace text representation of trees and tents with graphics~
* ~Add inputting aux. lines in auto mode~
* ~Run prettier~

### CODE REVIEW

File: src/variety/tents.js [hunk 1]
```diff
@@ -0,0 +1,499 @@
...
+			"(please translate) There aren't enough tents around a tree.",
+			"There aren't enough tents around a tree."
+		],
+		nmTentGt: [
+			"(please translate) There are too many tents around a tree.",
+			"There are too many tents around a tree."
+		],
+		tentAround: [
+			"(please translate) Tents are adjacent.",
+			"Tents are adjacent."
```
Comment (robx) [inline src/variety/tents.js]: I think (some) people read "adjacent" as touch by an edge. Would suggest "Some tents touch.", as we have with star battle.

File: src/variety/tents.js [hunk 2]
```diff
@@ -0,0 +1,499 @@
...
+	BoardExec: {
+		adjustBoardData: function(key, d) {
+			this.adjustEdge_1(key, d);
+		},
+		adjustBoardData2: function(key, d) {
+			this.adjustEdge_2(key, d);
+		}
+	},
+
+	// The rules state that there must be at least one valid matching between trees and tents.
```
Comment (robx) [inline src/variety/tents.js]: It's not clear to me that these are *the* rules for tents and trees. E.g., http://wiki.logic-masters.de/index.php?title=Zeltlager/de require an explicit matching, which would mean a 3x3 puzzle with four tents on the sides is ambiguous.

Not saying that the approach you're taking here is wrong, just that I'd still like to consider what to do about this.

Regardless of the rule interpretation, it would seem to be useful to be able to mark tree-tent connections.

**Comment by robx** [comment]
About the matching: I think the current implementation would work for both rulesets: If the rules require an explicit matching, that doesn't need to be reflected in the actual answer. (If we made the matching part of the answer the error messages could be more explicit, but I think that's something we could deal with.)

So my suggestion would be:
- Leave the answer checker as is.
- Add a variant rule "It is not required to match tents to trees uniquely." This is an entirely "meta" variant rule, in that it only determines whether a given puzzle is valid, or has multiple solutions. There may be better ways to express this.
- Add a way to input auxiliary marks to connect tents to trees.

I'm going with the explicitly matched variant as the default ruleset here, because I believe that to be more common and easier to explain. Happy to discuss though!

**Comment by robx** [comment]
I think I'd also prefer to leave out the marked empty cells (dash, qnum = 3, shade input mode). Any particular reason to support that?

**Comment by robx** [comment]
After some more discussion and research: The situation isn't clear-cut as to what's more common.

From my experience, in every Q&A session for contests with tents/trees where the rules aren't explicit on this matter, someone will ask for clarification. (Actually, they'll even do that when the rules *are* explicit.) I'd love to avoid that ambiguity.

However, I'm willing to consider that maybe this is not something we have to solve here. We might just stick with the current implementation and roll with the ambiguity. The primary effect that the difference in rules has is that in one case, you can use some uniqueness deductions which aren't safe in the other case; that's probably of minor importance to the majority of users of the solver.

Alternatively, if we're to have two variants, @edderiofer suggests:

[ when requiring a unique matching ] tell the player "there are multiple says to pair up tents and trees"

That is, add an answer check that errors an ambiguity for this variant.

**Comment by x-sheep** [comment]
On the graphics:

The only other puzzle with complex shapes is Goats and Wolves, which uses a base64-encoded PNG image. I don't think the Candle package contains a SVG deserializer. There also is no support for drawing ellipses, only circles and circular arcs.

I don't see a way to draw those tree graphics without making changes to the graphics library (unless there's a way to apply a scaling transform to just the trees...)

**Comment by edderiofer** [comment]
If puzz.link supports emoji, we could use a suitable tree emoji.

Alternatively, we could just use a different tree graphic. I don't particularly like the current one. And if we are going to use a PNG image, then we may as well colour it like a tree.

**Comment by robx** [comment]
Thanks for investigating! In that case it sounds like a PNG will be fine, the sheep and wolves never bothered me as low quality. (I thought they were probably SVG.)

I agree that some tasteful colored trees and tents could be nice. "No" to using emoji.

**Comment by x-sheep** [comment]
While the 🌳 emoji looks fine in a grid, the differences in art styles on each platform are massive. It would be better to have consistency across platforms.

If anyone wants to draw artwork for this puzzle, here's the file used by shwolf for reference:
shwolf

Note the lack of anti-aliasing, presumably to use as few colors as possible to keep the filesize down.

I think only a tree sprite is needed. The tent could be drawn from code, in any color.

**Comment by robx** [comment]
I'd prefer both to be sprites, for consistency. I'll ask around.

**Comment by robx** [comment]
I made some.

tent-tree

**Comment by x-sheep** [comment]
I've added the image after running it through pngcrush (to get it from 2034 to 432 bytes).

There currently isn't a way to differentiate between given tents and answer tents. Do we want to add a third image, or change the background for immutable cells?

**Comment by robx** [comment]
I don't think there should be given tents at all.

**Comment by robx** [comment]
Here's another option.

tent-tree

**Comment by robx** [comment]
Or this. A bit off-center. The geometry seems simple enough that we might be able to draw this in code, what do you think?

green-tent-tree

**Comment by robx** [comment]
The same in black and white might also work (or perhaps using some gray: fully black tree looks too heavy, while with white inside it doesn't look much like a tree)

**Comment by robx** [comment]
Some remarks based on testing the current version:

- removing tents should probably remove the connecting lines
- I'd like to be able to mark non-connection (probably a peke, although on paper I tend to draw a thick bar on the gridline)

**Comment by x-sheep** [comment]
I've added more cell background colors, mostly as an experiment. I've replaced the immutable dashes with shaded cells.

For the tent graphics, I think it's worth switching back to a vector drawing, just to deal with the various colors (clue, answer, error, trial, possibly mouse-down). I prefer the tents used in puzzle-draw over the later ones. I haven't decided on a tree graphic yet.

**Comment by x-sheep** [comment]
I've replaced all graphics with vectors.

**Comment by robx** [comment]
I like them! (Or the shape at least :)) Here's what they look like:

Screenshot 2020-02-29 at 15 27 19

Some remarks on the visuals:
- the tree colors are too dark; the trees would be better of solid black, or maybe with a light (white even?) interior
- could the tent color be made the same as the aux line color? (they shouldn't be close but different, and I think I'd prefer the darker tone)
- could the aux lines be drawn far enough to touch / go behind the tents when horizontal?

(Apart from that, I still don't like given tents or blanks.)

**Comment by x-sheep** [comment]
I don't know which other tree colors are visually distinct from the tent color, while still looking decent. A white or black fill looks bad IMO.

**Comment by robx** [comment]
I think it looks nice with some shade of lighter green filling and black:

Screenshot 2020-03-01 at 22 08 11

(This is `var crownfill = "rgb(212,251,121)";`.)

**Comment by robx** [comment]
Also I have to say it's pretty fun, just test-solved a couple puzzles, here's two due to Martin Merker from LM qualifiers 2018:

https://garp.vllmrt.net/scratch/tents/p.html?tents/10/10/h1g4k2i3kgbgbg1bgbgbqbnbgbhbgbnbqbgbg1bgbgb
https://garp.vllmrt.net/scratch/tents/p.html?tents/12/12/1j1h1i1j1i1hbhblbh6bs6bhbrbbbbhbbibkbs6bbhblbk6bhbb

Some remarks:

- It feels like you should be able to drag a tent over a dot; currently that's not possible.
- I think it would be nice on the other hand (but entirely optional) to make dragging dots not erase tents (similar to how right-click-dragging in heyawake leaves the shaded cells be); the typical usecase for dot-dragging seems to be marking empty around placed tents, and that's much easier if you can't accidentally lose them. (There's maybe the same issue here as with heyawake that this doesn't play nicely with one-button input and should be right-button only.)
- I'd still like to be able to place pekes to mark where a tent can't go from a tree.

**Comment by robx** [comment]
Just curious if this is done from your point of view? If so, I'd probably merge this after reworking to remove given blanks and trees.

**Comment by x-sheep** [comment]
Yes, I think it's done.

File: src/variety/tents.js [hunk 3]
```diff
@@ -0,0 +1,744 @@
...
+						} else {
+							this.inputData = this.btn === "left" ? 1 : 2;
+						}
+					}
+
+					if (hastree && hasdot) {
+						this.inputData = this.btn === "left" && !single ? 1 : 2;
+					}
+
+					this.inputcell_tents(other);
```
Comment (robx) [inline src/variety/tents.js]: I'm trying to understand why the starting cell is only input with delay. I guess this is the point where that happens?

E.g.:

1. left mouse down on cell (nothing happens)
2. drag to neighbour cell (tree appears in starting cell)
3. lift mouse button (nothing happens)

Similar behaviour occurs for dots. In both cases I'd expect the symbol to appear on first mouse-down, but I may be missing some reason why we can't know for sure that's going to happen at that point?

Comment (x-sheep) [reply src/variety/tents.js]: The delay is to differentiate between cell input and border input. Without a delay, you can't drag a line from an existing tent to a tree without erasing the tent. This is more noticable when you only want to remove a line without removing the tent. There's a lot of special edge-cases which are covered by the unit tests.

To get something to happen on mousedown, I recommend not actually changing the input order, but maybe set an additional value on the first cell (such as `info`) which allows for purely visual feedback. Mouseup would clear this value on all cells. You could use this to change the cell background to a special color like in Shugaku (School Trip), or to draw a tent when the actual cell is still empty.

**Comment by x-sheep** [comment]
I've incorporated the changes from #164, and changed the example puzzle according to https://github.com/robx/pzprjs/pull/164#issuecomment-598893431

File: src/variety/tents.js (hunk 3 above)
Comment (x-sheep) [reply src/variety/tents.js:99]: Right-clicking an empty cell now inputs a dot on mousedown, except when the mouse is on a border. Left-clicking an empty cell will display a fake tent on mousedown using `setinfo` as described above.

All actions still work the same as before, but now there's feedback on mousedown in most common cases.

**Comment by robx** [comment]
Thanks, this looks like it's ready to merge!

I've made one change, which is to use a more compact encoding, here: https://github.com/robx/pzprjs/tree/tents

I'll put the new version up for testing, then merge in a couple days if nothing comes up.

**Comment by robx** [comment]
Merged by hand, thanks!
